File: week6/sentimental-credit/credit.py
def main():
    print("Enter a card number: ", end="")
    number = input()  # getting user input
    while (not (number.isnumeric())):  # if not a number - request again
        print("Enter a card number: ", end="")
        number = input()
    firstNumbers = number[0] + number[1]  # leading digits
    if (luhn(number)):  # verifying by luhn's algorithm
        if (firstNumbers in ["34", "37"] and len(number) == 15):
            print("AMEX")
        elif (firstNumbers in ["51", "52", "53", "54", "55"] and len(number) == 16):
            print("MASTERCARD")
        elif (firstNumbers[0] == "4" and ((len(number) == 13) or (len(number) == 16))):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


def luhn(number):
    number = int(number)
    length = 0
    sum = 0
    while (number > 0):
        length += 1
        sum += number % 10
        number = int(number / 10)
        if (number > 0):
            length += 1
            tempProduct = (number % 10) * 2
            while (tempProduct > 0):
                sum += tempProduct % 10
                tempProduct = int(tempProduct / 10)
            number = int(number / 10)
    if (sum % 10 == 0):
        return True
    return False


main()

File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number = get_long("Enter a number: "); // getting input from user
    int sum = 0;
    int length = 0;
    int lastNumbers = 0; // sum by Luhn's Algorithm, long length, first two digits of number
    while (number > 0)
    {
        length++;
        if (number > 10)
            lastNumbers = number % 100; // getting last two numbers
        sum += number % 10;             // increasing sum by luhn's
        number = number / 10;           // decreasing the number
        if (number > 0)
        {
            length++;
            if (number > 10)
                lastNumbers = number % 100;
            int tempProduct = number % 10 * 2; // tempproduct might be >10
            while (tempProduct > 0)
            {
                sum += tempProduct % 10; // adding each of tempproduct digtits to sum
                tempProduct = tempProduct / 10;
            }
            number = number / 10; // decreasing the number
        }
    }
    if (sum % 10 == 0) // veryfying by Luhn's algorithm
    {
        if (length == 13) // veryfing by length and last numbers
        {
            if (lastNumbers / 10 == 4)
                printf("VISA\n");
            else
                printf("INVALID\n");
        }
        else if (length == 15)
        {
            if (lastNumbers == 34 || lastNumbers == 37)
                printf("AMEX\n");
            else
                printf("INVALID\n");
        }
        else if (length == 16)
        {
            if (lastNumbers == 51 || lastNumbers == 52 || lastNumbers == 53 || lastNumbers == 54 ||
                lastNumbers == 55)
                printf("MASTERCARD\n");
            else if (lastNumbers / 10 == 4)
                printf("VISA\n");
            else
                printf("INVALID\n");
        }
        else
            printf("INVALID\n");
    }
    else
        printf("INVALID\n");
}
